### PR TITLE
Simplify `Ed25519KeyPair` construction.

### PR DESCRIPTION
**Pull request by briansmith**
@ranweiler I had left a lot of comments on your PR that were annoying “make this simpler” type things. I got worried that maybe I was overlooking something, so I decided to play around with it a bit to make sure what I was asking for is reasonable. Could you check out these follow-ups to your PR and make sure they make sense? Thanks!

### CODE REVIEW

**Comment by ranweiler** [comment]
@briansmith Yes! I was just about to respond to that feedback, so I'm glad you opened this. I'll review it shortly.

**Comment by briansmith** [comment]
Part of this was motiviated by your suggestion of `check!()` to reduce `if`s, e.g. the `Simplify public key length` commits.

File: src/ec/eddsa.rs [hunk 1]
```diff
@@ -108,36 +86,28 @@ impl<'a> Ed25519KeyPair {
...
+        let private_key = try!(slice_as_array_ref!(private_key, SEED_LEN));
+        let public_key =
+            try!(slice_as_array_ref!(public_key, PUBLIC_KEY_LEN));
+
+        let pair = Ed25519KeyPair::from_bytes_unchecked(private_key, public_key);
         let mut public_key_check = [0; PUBLIC_KEY_LEN];
-        public_from_private(&pair.private_key, &mut public_key_check);
-        if public_key != public_key_check {
+        public_from_private(&private_key, &mut public_key_check);
+        if &public_key[..] != &public_key_check[..] {
```
Comment (ranweiler) [inline src/ec/eddsa.rs]: Is this change just to avoid the unnecessary move?

File: src/ec/eddsa.rs [hunk 2]
```diff
@@ -90,22 +86,28 @@ impl<'a> Ed25519KeyPair {
...
+        let pair = Ed25519KeyPair::from_seed(seed);
+        if &public_key[..] != &pair.public_key[..] {
             return Err(error::Unspecified);
         }
         Ok(pair)
     }
 
-    fn from_bytes_unchecked(seed: &Seed, public_key: &PublicKey)
-                            -> Ed25519KeyPair {
+    fn from_seed(seed: &Seed) -> Ed25519KeyPair {
```
Comment (ranweiler) [inline src/ec/eddsa.rs]: This change makes what's going on much clearer!

File: src/ec/eddsa.rs [hunk 3]
```diff
@@ -171,11 +171,8 @@ pub static ED25519: EdDSAParameters = EdDSAParameters {};
 impl signature::VerificationAlgorithm for EdDSAParameters {
     fn verify(&self, public_key: untrusted::Input, msg: untrusted::Input,
               signature: untrusted::Input) -> Result<(), error::Unspecified> {
-        if public_key.len() != PUBLIC_KEY_LEN {
-            return Err(error::Unspecified);
-        }
         let public_key = public_key.as_slice_less_safe();
-        let public_key = slice_as_array_ref!(public_key, ELEM_LEN).unwrap();
+        let public_key = try!(slice_as_array_ref!(public_key, ELEM_LEN));
```
Comment (ranweiler) [inline src/ec/eddsa.rs:175]: Hmm, this looks much more idiomatic. The only thing I'm hesitant about is that, per the `check!` discussion, the branching— and more importantly, input validation— now feels hidden.

I know that we're using the length check of `slice_as_array_ref!` to validate the input length. But, the existence of this check is less obvious now. I guess if we were only _really_ checking `public_key.len()` to ensure our `unwrap()` could never fail, then this is an unambiguous improvement. But, if we also care about `public_key.len()` being a valid length for its own sake, then I'd prefer not to push that check into the `slice_as_array_ref!` macro. In short, I'd prefer input validation to be explicit, and right now it feels a little implicit.

What do you think?

Comment (ranweiler) [reply src/ec/eddsa.rs:175]: Maybe we should just add a comment about the implicit check, like in the other "length" commit.

**Review by ranweiler — APPROVED** [review]
I really like how this replaced two private functions with a method that feels much more motivated (`from_seed`). These changes are a big improvement overall.

I think it would be best if all input validation were done in some very explicit way, which is one thing I like about either an `if {..}` or `check!`-style guard. We should think about how we want to approach that generally.

On the other hand, for the cases where it occurs here, the checks are minor enough (not "semantic", only length checks) where I think it is okay to do some of them implicitly, and simply note where we are leaning on something else to do the work.

**Comment by briansmith** [comment]
> Hmm, this looks much more idiomatic. The only thing I'm hesitant about is that, per the check! discussion, the branching— and more importantly, input validation— now feels hidden.

Yes, this is definitely an issue with using `try!()`. Unfortunately this issue is pervasive in the code base and in Rust code in general. We will need better tools that make sure that both branches of every `try!`/`?` are exercised. If/when we do that, then we might try the `check!()` thing again, if we can extend such tools to it. But I think the ship has sailed w.r.t. using `try!()` in any case where we return early do to a `Result::Err`.

**Comment by briansmith** [comment]
> I guess if we were only really checking public_key.len() to ensure our unwrap() could never fail, then this is an unambiguous improvement.

One thing to keep in mind is that we want to avoid `unwrap()`, and that's one reason I think that this is an improvement. IMO `unwrap()` is worse than `try!()`. We could write something like this:

```rust
let x = if let Ok(x) = whatever() {
    x
} else {
    return Err(error::Unspecified);
}
```

But that's just horrendously verbose and ugly.

**Comment by briansmith** [comment]
> I think it would be best if all input validation were done in some very explicit way, which is one thing I like about either an if {..} or check!-style guard. We should think about how we want to approach that generally.

Generally I do want checks to be explicit. I think `try!()` is explicit. Redundant checks (which would end up with dead code branches) or `unwrap()` are worse. So, I would say the rules are something like:

0. Make sure all the checks get done.
1. Prefer `try!()` whenever it makes sense; we'll deal with the negative effects of using it later.
2. Don't introduce dead code paths, especially with redundant checks.
3. If a check does two things, one of which is non-obvious, document the non-obvious effect of the check.
4. Avoid unwrap().
5. Sometimes efficiency concerns or other concerns might override the above; in those cases, document the motivation for the divergence from these guidelines.

File: src/ec/eddsa.rs (hunk 1 above)
Comment (briansmith) [reply src/ec/eddsa.rs]: `[u8; 32]` doesn't have operator `!=` implemented for it. Basically this is just a workaround for the lack of generics parameterized by array length.

**Comment by briansmith** [comment]
> Maybe we should just add a comment about the implicit check, like in the other "length" commit.

I think if `try!(slice_as_array_ref!(...))` is unclear then we should fix it to make it clear, because it is designed to be used exactly that way, and that's how it is used throughout the code base. It's good to know that somebody finds it be be unclear.

**Comment by briansmith** [comment]
Thanks again!

**Pull request closed without merging**